File: Assets/Scripts/AI/latency_test.py
#!/usr/bin/env python3
"""
Yazh 30K ONNX Model — Latency Validation Script
Runs inference on the ONNX model using ONNX Runtime and measures latency.
Can be run independently of Unity.

Usage:
    python3 latency_test.py [--model MODEL_PATH] [--iterations N] [--warmup N]
"""

import argparse
import json
import os
import sys
import time
import statistics
from pathlib import Path

def check_onnxruntime():
    """Check if onnxruntime is installed."""
    try:
        import onnxruntime as ort
        print(f"[OK] onnxruntime {ort.__version__}")
        return True
    except ImportError:
        print("[WARN] onnxruntime not installed. Install with: pip install onnxruntime")
        print("[INFO] Falling back to ONNX file validation only.")
        return False

def validate_model(model_path):
    """Validate the ONNX model structure without running inference."""
    try:
        import onnx
        model = onnx.load(model_path)
        onnx.checker.check_model(model)

        print(f"[OK] Model valid: {model_path}")
        print(f"  IR version: {model.ir_version}")
        opsets = [f"{o.domain or 'ai.onnx'}:{o.version}" for o in model.opset_import]
        print(f"  Opset: {opsets}")
        print(f"  Graph name: {model.graph.name}")
        print(f"  Inputs:")
        for inp in model.graph.input:
            shape = [d.dim_value if d.dim_value else d.dim_param for d in inp.type.tensor_type.shape.dim]
            dtype = inp.type.tensor_type.elem_type
            print(f"    - {inp.name}: shape={shape}, dtype={dtype}")
        print(f"  Outputs:")
        for out in model.graph.output:
            shape = [d.dim_value if d.dim_value else d.dim_param for d in out.type.tensor_type.shape.dim]
            dtype = out.type.tensor_type.elem_type
            print(f"    - {out.name}: shape={shape}, dtype={dtype}")

        # Count parameters
        total_params = 0
        for initializer in model.graph.initializer:
            params = 1
            for d in initializer.dims:
                params *= d
            total_params += params
        print(f"  Total parameters: {total_params:,} ({total_params / 1e6:.1f}M)")

        # Model size
        size_mb = os.path.getsize(model_path) / (1024 * 1024)
        print(f"  File size: {size_mb:.1f} MB")

        return True
    except ImportError:
        print("[WARN] onnx package not installed. Skipping structural validation.")
        size_mb = os.path.getsize(model_path) / (1024 * 1024) if os.path.exists(model_path) else 0
        print(f"  File exists: {os.path.exists(model_path)}, Size: {size_mb:.1f} MB")
        return True
    except Exception as e:
        print(f"[ERROR] Model validation failed: {e}")
        return False

def run_inference_benchmark(model_path, iterations=20, warmup=3):
    """Run inference benchmark using ONNX Runtime."""
    import onnxruntime as ort
    import numpy as np

    print(f"\n{'='*50}")
    print(f"  YAZH 30K LATENCY BENCHMARK")
    print(f"{'='*50}")
    print(f"  Model: {model_path}")
    print(f"  Iterations: {iterations}, Warmup: {warmup}")

    # Create session
    sess_options = ort.SessionOptions()
    sess_options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
    sess_options.intra_op_num_threads = 4

    try:
        session = ort.InferenceSession(model_path, sess_options)
    except Exception as e:
        print(f"[ERROR] Failed to create session: {e}")
        return None

    # Get input info
    input_name = session.get_inputs()[0].name
    input_shape = session.get_inputs()[0].shape
    input_dtype = session.get_inputs()[0].type
    print(f"  Input: {input_name} shape={input_shape} dtype={input_dtype}")

    output_name = session.get_outputs()[0].name
    output_shape = session.get_outputs()[0].shape
    print(f"  Output: {output_name} shape={output_shape}")

    seq_len = input_shape[1] if len(input_shape) > 1 else 256

    # Warmup
    print(f"\n  [WARMUP] {warmup} iterations...")
    dummy_input = np.random.randint(0, 30000, size=(1, seq_len)).astype(np.int64)
    for i in range(warmup):
        try:
            session.run(None, {input_name: dummy_input})
            print(f"    Warmup {i+1}/{warmup} — OK")
        except Exception as e:
            print(f"    Warmup {i+1}/{warmup} — FAILED: {e}")
            return None

    # Benchmark
    print(f"\n  [BENCHMARK] {iterations} iterations...")
    latencies = []
    success = 0

    for i in range(iterations):
        # Random Tamil-like token IDs
        test_input = np.random.randint(0, 30000, size=(1, seq_len)).astype(np.int64)

        start = time.perf_counter()
        try:
            output = session.run(None, {input_name: test_input})
            end = time.perf_counter()
            latency_ms = (end - start) * 1000
            latencies.append(latency_ms)
            success += 1
            print(f"    [{i+1:02d}/{iterations}] {latency_ms:8.2f} ms")
        except Exception as e:
            print(f"    [{i+1:02d}/{iterations}] FAILED: {e}")

    # Statistics
    if latencies:
        latencies.sort()
        avg = statistics.mean(latencies)
        med = statistics.median(latencies)
        std = statistics.stdev(latencies) if len(latencies) > 1 else 0
        p50 = latencies[int(len(latencies) * 0.50)]
        p90 = latencies[int(len(latencies) * 0.90)]
        p95 = latencies[int(len(latencies) * 0.95)]
        min_l = min(latencies)
        max_l = max(latencies)

        passed = avg < 150
        verdict = "PASS" if passed else "FAIL"

        print(f"\n{'='*50}")
        print(f"  RESULTS")
        print(f"{'='*50}")
        print(f"  Success:     {success}/{iterations}")
        print(f"  Min:         {min_l:.2f} ms")
        print(f"  Avg:         {avg:.2f} ms")
        print(f"  Median:      {med:.2f} ms")
        print(f"  P50:         {p50:.2f} ms")
        print(f"  P90:         {p90:.2f} ms")
        print(f"  P95:         {p95:.2f} ms")
        print(f"  Max:         {max_l:.2f} ms")
        print(f"  Std Dev:     {std:.2f} ms")
        print(f"  Target:      < 150 ms")
        print(f"  Verdict:     {verdict}")
        print(f"{'='*50}")

        # Save results
        results = {
            "model": model_path,
            "timestamp": time.strftime("%Y-%m-%d %H:%M:%S"),
            "iterations": iterations,
            "success_count": success,
            "latency_ms": {
                "min": round(min_l, 2),
                "avg": round(avg, 2),
                "median": round(med, 2),
                "p50": round(p50, 2),
                "p90": round(p90, 2),
                "p95": round(p95, 2),
                "max": round(max_l, 2),
                "std": round(std, 2)
            },
            "target_ms": 150,
            "passed": passed,
            "raw_latencies": [round(l, 2) for l in latencies]
        }

        results_path = Path(model_path).parent / "latency_results.json"
        with open(results_path, 'w') as f:
            json.dump(results, f, indent=2)
        print(f"\n  [SAVE] Results saved to: {results_path}")

        return results
    else:
        print("\n  [ERROR] No successful inferences.")
        return None

def main():
    parser = argparse.ArgumentParser(description="Yazh 30K ONNX Latency Test")
    parser.add_argument("--model", type=str,
                        default="MLModels/yazh-30k-int8.onnx",
                        help="Path to ONNX model (relative to StreamingAssets)")
    parser.add_argument("--iterations", type=int, default=20,
                        help="Number of benchmark iterations")
    parser.add_argument("--warmup", type=int, default=3,
                        help="Number of warmup iterations")
    parser.add_argument("--streaming-assets", type=str,
                        default="/home/neutron/Yazhi/apps/yazh-unity/Assets/StreamingAssets",
                        help="Path to StreamingAssets directory")
    args = parser.parse_args()

    # Resolve model path
    model_path = os.path.join(args.streaming_assets, args.model)
    if not os.path.exists(model_path):
        print(f"[ERROR] Model not found: {model_path}")
        sys.exit(1)

    print(f"Model: {model_path}")
    print(f"Size: {os.path.getsize(model_path) / (1024*1024):.1f} MB")

    # Step 1: Validate model structure
    print(f"\n--- Step 1: Model Validation ---")
    if not validate_model(model_path):
        sys.exit(1)

    # Step 2: Run inference benchmark
    print(f"\n--- Step 2: Inference Benchmark ---")
    has_ort = check_onnxruntime()
    if has_ort:
        results = run_inference_benchmark(model_path, args.iterations, args.warmup)
        if results is None:
            sys.exit(1)
    else:
        print("[SKIP] Cannot run inference without onnxruntime.")
        print("[INFO] The Unity LatencyTest.cs component will run the benchmark in-editor.")

    print("\n[DONE] Latency test complete.")

if __name__ == "__main__":
    main()

File: Assets/Scripts/AI/LatencyTest.cs
// LatencyTest.cs — YAZH-UNITY
// ARIVU | Rotation 25 | Jun 17, 2026
// Benchmarks Yazh 30K ONNX model inference latency on target hardware

using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using System.IO;

public class LatencyTest : MonoBehaviour
{
    [Header("Test Configuration")]
    public int warmupIterations = 5;
    public int benchmarkIterations = 20; // Clamped to [1, 100]
    public string testInput = "வணக்கம்"; // "Hello" in Tamil

    [Header("Output")]
    public string resultsPath = "latency_results.json";

    private YazhInferenceManager inference;
    private List<float> latencies = new List<float>();
    private bool testRunning = false;

    void Start()
    {
        inference = YazhInferenceManager.Instance;
        if (inference == null)
        {
            Debug.LogError("[LatencyTest] YazhInferenceManager not found!");
            enabled = false;
            return;
        }
    }

    /// <summary>
    /// Run the full latency benchmark. Call from another script or UI button.
    /// </summary>
    public void RunBenchmark()
    {
        if (testRunning)
        {
            Debug.LogWarning("[LatencyTest] Test already running!");
            return;
        }
        StartCoroutine(BenchmarkCoroutine());
    }

    IEnumerator BenchmarkCoroutine()
    {
        testRunning = true;
        latencies.Clear();

        // Validate iteration count
        benchmarkIterations = Mathf.Clamp(benchmarkIterations, 1, 100);

        Debug.Log("[LatencyTest] Starting benchmark — warmup: " + warmupIterations +
                  ", iterations: " + benchmarkIterations);

        // Wait for model to be ready
        while (!inference.IsModelReady())
        {
            Debug.Log("[LatencyTest] Waiting for model...");
            yield return new WaitForSeconds(1f);
        }

        // Warmup phase
        Debug.Log("[LatencyTest] Warmup phase...");
        for (int i = 0; i < warmupIterations; i++)
        {
            bool done = false;
            inference.GenerateResponse(testInput, (result) => { done = true; });
            while (!done) yield return null;
            yield return new WaitForSeconds(0.1f);
        }

        // Benchmark phase
        Debug.Log("[LatencyTest] Benchmark phase...");
        for (int i = 0; i < benchmarkIterations; i++)
        {
            float startTime = Time.realtimeSinceStartup;
            bool done = false;

            inference.GenerateResponse(testInput, (result) => {
                done = true;
            });

            while (!done) yield return null;

            float elapsed = (Time.realtimeSinceStartup - startTime) * 1000f; // ms
            latencies.Add(elapsed);

            Debug.Log("[LatencyTest] Iteration " + (i + 1) + ": " + elapsed.ToString("F1") + "ms");
            yield return new WaitForSeconds(0.05f); // Small gap between iterations
        }

        // Calculate results
        float min = float.MaxValue, max = 0, sum = 0;
        foreach (float l in latencies)
        {
            if (l < min) min = l;
            if (l > max) max = l;
            sum += l;
        }
        float avg = sum / latencies.Count;

        // Calculate median
        List<float> sorted = new List<float>(latencies);
        sorted.Sort();
        float median = sorted[sorted.Count / 2];

        // Calculate p95
        int p95Index = Mathf.FloorToInt(sorted.Count * 0.95f);
        float p95 = sorted[Mathf.Clamp(p95Index, 0, sorted.Count - 1)];

        // Target check
        bool meetsTarget = avg < 150f; // 150ms target

        // Build results JSON
        string json = "{\n";
        json += "  \"timestamp\": \"" + System.DateTime.UtcNow.ToString("o") + "\",\n";
        json += "  \"model\": \"Yazh 30K\",\n";
        json += "  \"test_input\": \"" + testInput + "\",\n";
        json += "  \"iterations\": " + latencies.Count + ",\n";
        json += "  \"results_ms\": {\n";
        json += "    \"min\": " + min.ToString("F2") + ",\n";
        json += "    \"max\": " + max.ToString("F2") + ",\n";
        json += "    \"avg\": " + avg.ToString("F2") + ",\n";
        json += "    \"median\": " + median.ToString("F2") + ",\n";
        json += "    \"p95\": " + p95.ToString("F2") + "\n";
        json += "  },\n";
        json += "  \"target_ms\": 150,\n";
        json += "  \"meets_target\": " + meetsTarget.ToString().ToLower() + ",\n";
        json += "  \"all_latencies\": [";
        for (int i = 0; i < latencies.Count; i++)
        {
            json += latencies[i].ToString("F2");
            if (i < latencies.Count - 1) json += ", ";
        }
        json += "]\n";
        json += "}";

        // Save results
        string path = Path.Combine(Application.persistentDataPath, resultsPath);
        File.WriteAllText(path, json);

        // Log summary
        Debug.Log("=== LATENCY TEST RESULTS ===");
        Debug.Log("Model: Yazh 30K");
        Debug.Log("Iterations: " + latencies.Count);
        Debug.Log("Min: " + min.ToString("F1") + "ms");
        Debug.Log("Max: " + max.ToString("F1") + "ms");
        Debug.Log("Avg: " + avg.ToString("F1") + "ms");
        Debug.Log("Median: " + median.ToString("F1") + "ms");
        Debug.Log("P95: " + p95.ToString("F1") + "ms");
        Debug.Log("Target (<150ms): " + (meetsTarget ? "PASS ✓" : "FAIL ✗"));
        Debug.Log("Results saved to: " + path);
        Debug.Log("===========================");

        testRunning = false;
    }

    /// <summary>
    /// Quick single inference test. Returns latency in ms.
    /// </summary>
    public IEnumerator QuickTest()
    {
        if (inference == null || !inference.IsModelReady())
        {
            Debug.LogError("[LatencyTest] Model not ready!");
            yield break;
        }

        float start = Time.realtimeSinceStartup;
        bool done = false;
        string response = "";

        inference.GenerateResponse(testInput, (result) => {
            response = result;
            done = true;
        });

        while (!done) yield return null;

        float latency = (Time.realtimeSinceStartup - start) * 1000f;
        Debug.Log("[LatencyTest] Quick test: " + latency.ToString("F1") +
                  "ms | Response: " + response);
    }
}
